### PR TITLE
Add Antimatter Power Generation

### PR DESCRIPTION
**Pull request by S4mpsa**
Adds a new endgame power option, **Semi-Stamble Antimatter**, for UIV, UMV and UXV power generation. Intended to tackle the reliance on Dysons in UIV and providing an alternative to EOH power generation (while still keeping EOHs the best power generation option overall).

Antimatter power consists of two new multiblocks, the Semi-Stable Antimatter Stabilization Sequencer (Antimatter Forge), and the Shielded Lagrangian Annihilation Matrix (Antimatter Generator).

### Antimatter Forge

image

image

The Forge is used to convert Protomatter, a new fluid, into Antimatter. The forge has an unique automation challenge based on balancing the antimatter amount in 16 hatches, and optimizing the amount of antimatter with energy. The production of Antimatter can be upgraded with different fluids to make it scale to UMV and UXV, while introducing logistical complexity to sustaining Antimatter power.

The balancing math for Antimatter is quite complex, see the Sheet here: 
[AntimatterMath.xlsx](https://github.com/user-attachments/files/16922254/AntimatterMath.xlsx)

TL;DR
2.77T in UIV, 23.2T in UMV, 310T in UXV of net EU/t in Antimatter.
After running through generator:
3.84T in UIV, 40,8T in UMV, 800T in UXV

To reach these numbers, a large amount of tiered materials need to be supplied to the Forge and the Generator.
The energy values do not account for the energy lost for the creation of the booster fluids, and on average the player will have to spend 10-30% of the energy produced just to sustain the antimatter production.

### Antimatter Generator

image

image

The generator converts Antimatter and Matter directly into EU. It can process infinite EU per operation, but is capped by the amount of source hatches. Varying Matter can be used to increase the EU of Antimatter in an exponential fashion, providing an incentive for the player to annihilate as much Antimatter as they can in a single cycle, without saturating the energy hatches.

Also supports wireless mode in UMV, but still respects the hatches.

This PR contains the base mechanics of Antimatter with fully functional machines, but is lacking recipes for protomatter and the machines themselves. Some rendering and texture work is still missing, but the multis should be in a state where they can be already tested.

Thanks to @FourIsTheNumber for help with the generator, @CookieBrigade for the rendering of the Forge and @BlueWeabo for help with some of the code.
The structure was jointly drafted between @S4mpsa, @NotAPenguin0 and Deleno.

### CODE REVIEW

File: src/main/java/goodgenerator/blocks/tileEntity/AntimatterGenerator.java [hunk 1]
```diff
@@ -0,0 +1,505 @@
...
+
+    @Override
+    public RecipeMap<?> getRecipeMap() {
+        return null;
+    }
+
+    @Override
+    public String[] getInfoData() {
+        String[] info = super.getInfoData();
+        info[4] = "Probably makes: " + EnumChatFormatting.RED
```
Comment (NotAPenguin0) [inline src/main/java/goodgenerator/blocks/tileEntity/AntimatterGenerator.java:152]: nit: I think this is usually worded as `Probably generates...` in other generators.

```suggestion
        info[4] = "Probably generates: " + EnumChatFormatting.RED
```

Comment (S4mpsa) [reply src/main/java/goodgenerator/blocks/tileEntity/AntimatterGenerator.java:152]: Need to redo this function entirely anyways, generator works in bursts instead of continously.

**Comment by S4mpsa** [comment]
There's still some polish required (Scanner info and GUI). Shouldn't merge into main pack before I do those, but everything should work mechanics wise for testing in nightlies.

**Comment by Dream-Master** [comment]
@S4mpsa any news here? Want to merge this soon. Will focus on beta version this weekend

File: src/main/java/goodgenerator/blocks/tileEntity/AntimatterGenerator.java [hunk 2]
```diff
@@ -0,0 +1,505 @@
...
+        long generatedEU = 0;
+
+        if (modifier != null) {
+            generatedEU = (long) ((Math.pow(antimatter, modifier) * 1e12) * (Math
+                .min(((float) antimatter / (float) catalystCount), ((float) catalystCount / (float) antimatter))));
+        }
+
+        if (wirelessEnabled && modifier >= 1.03F) {
+            // Clamp the EU to the maximum of the hatches so wireless cannot bypass the limitations
+            long euCapacity = 0;
```
Comment (NotAPenguin0) [inline src/main/java/goodgenerator/blocks/tileEntity/AntimatterGenerator.java:215]: This can probably be computed once on structure check in `checkMachine` and then be cached

File: src/main/java/goodgenerator/blocks/tileEntity/AntimatterGenerator.java [hunk 3]
```diff
@@ -0,0 +1,505 @@
...
+    }
+
+    @Override
+    public void construct(ItemStack itemStack, boolean hintsOnly) {
+        buildPiece(MAIN_NAME, itemStack, hintsOnly, 17, 41, 0);
+    }
+
+    @Override
+    public void saveNBTData(NBTTagCompound nbt) {
+        nbt = (nbt == null) ? new NBTTagCompound() : nbt;
```
Comment (NotAPenguin0) [inline src/main/java/goodgenerator/blocks/tileEntity/AntimatterGenerator.java:251]: If the passed in `nbt` is `null` then saving to this new instance won't do anything anyway, so it being null wouldn't make much sense. I'm pretty sure this is never `null`, so that should probably be removed.

File: src/main/java/goodgenerator/blocks/tileEntity/AntimatterGenerator.java [hunk 4]
```diff
@@ -0,0 +1,505 @@
...
+        nbt.setBoolean("wirelessEnabled", wirelessEnabled);
+        super.saveNBTData(nbt);
+    }
+
+    @Override
+    public void loadNBTData(NBTTagCompound nbt) {
+        nbt = (nbt == null) ? new NBTTagCompound() : nbt;
```
Comment (NotAPenguin0) [inline src/main/java/goodgenerator/blocks/tileEntity/AntimatterGenerator.java:258]: I don't think this can ever be null, but if it is you should just `return` instead of loading from an empty NBT tag

File: src/main/java/goodgenerator/blocks/tileEntity/AntimatterForge.java [hunk 5]
```diff
@@ -0,0 +1,961 @@
...
+    }
+
+    public int getCasingMeta(int type) {
+        switch (type) {
+            case 1:
+                return 0;
+            case 2:
+                return 0;
+            default:
+                return 0;
```
Comment (NotAPenguin0) [inline src/main/java/goodgenerator/blocks/tileEntity/AntimatterForge.java:352]: :caught:

File: src/main/java/goodgenerator/blocks/tileEntity/AntimatterForge.java [hunk 6]
```diff
@@ -0,0 +1,961 @@
...
+    }
+
+    public int getFrameMeta() {
+        return 0;
+    }
+
+    public int textureIndex(int type) {
+        switch (type) {
+            case 1:
+                return (12 << 7) + 9;
```
Comment (NotAPenguin0) [inline src/main/java/goodgenerator/blocks/tileEntity/AntimatterForge.java:387]: Probably better to look these up from the block class instead of hardcoding

File: src/main/java/goodgenerator/blocks/tileEntity/AntimatterForge.java [hunk 7]
```diff
@@ -0,0 +1,961 @@
...
+        }
+
+        List<FluidStack> inputFluids = getStoredFluids();
+        for (int i = 0; i < inputFluids.size(); i++) {
+            FluidStack inputFluid = inputFluids.get(i);
+            if (inputFluid.isFluidEqual(MaterialsUEVplus.Protomatter.getFluid(1))) {
+                containedProtomatter += inputFluid.amount;
+                continue;
+            }
+            for (int tier = 1; tier <= magneticUpgrades.length; tier++) {
```
Comment (NotAPenguin0) [inline src/main/java/goodgenerator/blocks/tileEntity/AntimatterForge.java:538]: This logic can definitely be abstracted into a function instead of copy pasting it 4 times

File: src/main/java/goodgenerator/blocks/tileEntity/AntimatterForge.java [hunk 8]
```diff
@@ -0,0 +1,961 @@
...
+    private static final float activeBaseExp = 1.5f;
+    private static final float coefficientBaseExp = 0.5f;
+    private static final float baseSkew = 0.2f;
+
+    private float[] modifiers = { 0.0f, 0.0f, 0.0f, 0.0f };
+    private FluidStack[] upgradeFluids = { null, null, null, null };
+    private int[] fluidConsumptions = { 0, 0, 0, 0 };
+
+    public static final String MAIN_NAME = "antimatterForge";
+    public static final int M = 1_000_000;
```
Comment (NotAPenguin0) [inline src/main/java/goodgenerator/blocks/tileEntity/AntimatterForge.java:95]: This should probably be renamed to not conflict/be confused with `GTValues.M`

**Review by NotAPenguin0 — CHANGES_REQUESTED** [review]

**Review by NotAPenguin0 — APPROVED** [review]
